carma_wm/include/carma_wm/WorldModel.h

```diff
@@ -0,0 +1,282 @@
+#include <lanelet2_routing/RoutingGraph.h>
+
+namespace carma_wm
+{
+  // Helpful using declarations which are not defined by lanelet::routing module
+  using LaneletRoutePtr = std::shared_ptr<lanelet::routing::Route>;
+  using LaneletRouteConstPtr = std::shared_ptr<const lanelet::routing::Route>;
+  using LaneletRouteUPtr = std::unique_ptr<lanelet::routing::Route>;
+  using LaneletRouteUConstPtr = std::unique_ptr<const lanelet::routing::Route>;
+
+  using LaneletRoutingGraphPtr = std::shared_ptr<lanelet::routing::RoutingGraph>;
+  using LaneletRoutingGraphConstPtr = std::shared_ptr<const lanelet::routing::RoutingGraph>;
+  using LaneletRoutingGraphUPtr = std::unique_ptr<lanelet::routing::RoutingGraph>;
+  using LaneletRoutingGraphConstUPtr = std::unique_ptr<const lanelet::routing::RoutingGraph>;
+
+  /*! \brief Position in a track based coordinate system where the axis are downtrack and crosstrack. 
+   *         Positive crosstrack is to the left of the reference line
+   *
+   * The position of a point relative to a reference line. The perpendicular distance from the point to the reference line is the crosstrack
+   * The distance to the intersection of the perpendicular from the start of the line is the downtrack distance
+   * If an x,y point would be located to the right of the reference line then the crosstrack is positive. It is negative if on the left side of the line.
+   * NOTE: The positive/negative crosstrack side is flipped relative to a lanelet::ArcCoordinate
+   * Most utility functions will assume that the downtrack/crosstrack values are specified in meters
+   */
+  class TrackPos
+  {
+    public:
+      double downtrack = 0;
+      double crosstrack = 0;
+
+      /*! \brief Constructor
+        * 
+        * \param down_track The downtrack distance
+        * \param cross_track The crosstrack distance
+        */
+      TrackPos(double down_track, double cross_track) 
+      : downtrack(down_track), crosstrack(cross_track) {}
+      /*! \brief Constructor from lanelet ArcCoordinates 
+        *         which are converted to TrackPos where downtrack = ArcCoordinates.length and crosstrack = -ArcCoordinates.distance
+        * 
+        * \param arc_coord ArcCoordinate to copy
+        */
+      TrackPos(const lanelet::ArcCoordinates& arc_coord) 
+      : downtrack(arc_coord.length), crosstrack(-arc_coord.distance) {}
+
+      /*! \brief Returns a lanelet ArcCoordinate built from this TrackPos
+        *        where downtrack = ArcCoordinates.length and crosstrack = -ArcCoordinates.distance
+        * 
+        * \return lanelet::ArcCoordinates with values copied from this object
+        */
+      inline lanelet::ArcCoordinates toArcCoordinates() {
+        return lanelet::ArcCoordinates{downtrack, -crosstrack};
+      }
+      
+      // Overload == and != operators
+      bool operator==(const TrackPos &other) const {
+        return this == &other 
+          || (this->downtrack == other.downtrack
```

Comment:
Does this really make sense as a raw double comparison? Should we consider adding an epsilon factor here to ensure some sanity?

Reply:
I believe in this case it should be a direct comparison for ==. This way a copy constructed object will evaluate as equal but a computed one will need the user to account for epsilons themselves and therefore be aware of the desired range.